### PR TITLE
perf(sessds): rotate through streams with iterators when fetching

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-12611](https://emqx.atlassian.net/browse/EMQX-12611).

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] ~~Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~~
- [x] Schema changes are backward compatible

[EMQX-12611]: https://emqx.atlassian.net/browse/EMQX-12611?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl [hunk 1]
```diff
@@ -534,6 +544,12 @@ n_awaiting_rel(Rec) ->
 
 %%
 
+-spec iterate(iter(K, V)) -> {K, V, iter(K, V)} | none.
+iterate(It0) ->
```
Comment (thalesmg) [inline apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl:548]: Tiny nit: maybe:

```suggestion
next(It0) ->
```

?

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl [hunk 2]
```diff
@@ -670,13 +694,110 @@ pmap_commit(
...
+            gbt_fold_iter(Fun, Fun(K, V, Acc), It);
+        _ ->
+            Acc
+    end.
+
+gbt_size(Cache) ->
+    gb_trees:size(Cache).
+
+gbt_iter_after(After, Cache) ->
+    It0 = gb_trees:iterator_from(After, Cache),
```
Comment (thalesmg) [inline apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl:790]: Q: does `beginning` happen to work due to the Erlang term order and due to the particular keys we use here?  Maybe we should have an explicit clause for `beginning`?

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_stream_scheduler.erl [hunk 3]
```diff
@@ -87,23 +111,41 @@ find_new_streams(S) ->
...
+            case Filter(Key, Stream) of
+                true ->
+                    {Key, Stream, ItStream};
+                false ->
+                    next_stream(ItStream)
+            end;
+        none ->
+            %% Restart the iteration from the beginning:
+            ItStream = ItStream0#iter{it = ItCont},
+            next_stream(ItStream)
```
Comment (thalesmg) [inline apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_stream_scheduler.erl:142]: Shouldn't we decrement `limit` before recursing here?

**Review by ieQu1 — APPROVED** [review]

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl (hunk 1 above)
Comment (ieQu1) [reply apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl:548]: Yes, I guess this name is better for the sake of naming consistency. We use `next` everywhere else.

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl:790]: Yes, the term order helps. I will leave a comment here to describe that we rely on it here.

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_stream_scheduler.erl (hunk 3 above)
Comment (keynslug) [reply apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_stream_scheduler.erl:142]: Good point. We haven't gone through one more element, thus no decrement. The situation you're describing can happen in general, but should not here: `it_cont` returning `none` right away means that `limit` is `0`. But it still sounds like a good idea to defend from that honestly.

**Review by ieQu1 — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3412)

**Pull request merged**